[WIP] -MFCS- Modular Firearm Construction System

BlueNexus opened the pull request:
//THIS IS HEAVILY WORK IN PROGRESS//

The MFCS allows players to construct and disassemble firearms using a wide variety of components, allowing the player to customise recoil, accuracy, projectile type, fire rate and mode, projectile speed, loading method and more. All firearms in the game will be replaced with these modular variants. MFCS Firearms may be constructed by R&D or anybody who gets a hold of the parts, however the final part will come in a lock box, to prevent them being made and used for grief. This will hopefully give research something interesting to do, and will add a lot of depth to the firearms in the game. Keep in mind, this project is massively work in progress. Feedback would be appreciated, however.

Current build - 215

BlueNexus committed:
Added the framework files for components

BlueNexus committed:
Added the different chambers

BlueNexus committed:
WIP Modular firearm code

Lorwp commented:
Is this in a state where it could be tested?

BlueNexus commented:
@Lorwp if you mean does it compile, it does. However, all that's finished is the first two steps of construction, and there's currently a fatal infinite loop error when attempting the second step.

BlueNexus committed:
Changed allowed_projectiles to use caliber

So much cleaner.
Also gave some things a placeholder sprite

BlueNexus committed:
Removed a stray piece of text

BlueNexus committed:
Moved projectile type into chamber/

BlueNexus committed:
Placeholder code for drivers

Lorwp commented:
@BlueNexus Got it, keep us posted on this, it looks interesting :+1:

BlueNexus committed:
Placeholder drivers to work with

BlueNexus committed:
Assembly driver code.

Heavily WIP and untested.

BlueNexus committed:
Adds a debug check

BlueNexus committed:
Duplicate definition fix

BlueNexus committed:
Reworked driver firemodes

BlueNexus committed:
Reworked driver firemodes

BlueNexus committed:
Added loaders

BlueNexus committed:
Moved the C.shell under shell

BlueNexus committed:
Added loaders, and commenting

BlueNexus commented:
Update - Stage four of the assembly is complete. 
Drivers = Done
Loaders = Done
Chassis = Sort of done
Barrels = Not started
Chambers = Done
Accessories = Not started

BlueNexus committed:
Quick patch

BlueNexus committed:
Will test travis

BlueNexus committed:
Revert

BlueNexus committed:
Create energy.dm

BlueNexus committed:
Create automatic

BlueNexus committed:
Create shotgun.dm

BlueNexus committed:
Rename automatic to automatic.dm

BlueNexus committed:
Create sniper.dm

BlueNexus committed:
Create pistol.dm

BlueNexus committed:
Delete energy.dm

BlueNexus committed:
Create beam.dm

BlueNexus committed:
Create special.dm

BlueNexus committed:
Create nuclear.dm

BlueNexus committed:
Create dart.dm

BlueNexus committed:
Added energy loaders

BlueNexus committed:
Var refactor

Will now use a compile_gun() proc to load all the vars, so that premade guns work

BlueNexus committed:
Removed obsolete code

BlueNexus committed:
Wiped core.dm, different purpose

BlueNexus committed:
Continue fix

BlueNexus committed:
Var fix

faerindel commented on code/modules/projectiles/guns/modular/assembly.dm:
```diff
@@ -0,0 +1,134 @@
+			I.loc = src
+			components += I
+			modLoader = I
+			var/obj.item/weapon/modular_firearms/loader/L = I
```
Typo? Or Byond swallows that var/obj.item?

BlueNexus commented:
@faerindel _sighs_ Typo. Thanks for pointing it out.

BlueNexus committed:
Major component refactor

A proc is now used, rather than repeating variables for each time an object is used on the frame. This also means that the proc can be called without attacking the frame, such as if an admin triggers it. This will also be used for dissasembly in the future.

BlueNexus committed:
Further refactor

Super-clean way of doing it, removing dozens of unnecessary lines of code. This is so CLEEEEEEEEEAN

BlueNexus committed:
Removed unnecessary vars

Buildstage has been entirely removed, thanks to the refactor.

BlueNexus committed:
Cleaning up code

BlueNexus committed:
Moved procs into core

BlueNexus committed:
Moved procs into core

BlueNexus commented:
Cleaned it up to massively reduce the amount of code needed. Adding parts and processing them is now handled by a single proc and a sub-proc. Will finish the current parts, then work on deconstruction. Finally, processing the gun itself.

BlueNexus committed:
Added testing SRC's

Not sure if these are needed, worth seeing if they fix the compile errors.

BlueNexus committed:
Added barrels

BlueNexus committed:
Added some +fun barrels

Good luck ever getting them though.

BlueNexus committed:
Added barrel handling to core

BlueNexus committed:
Added stocks

BlueNexus committed:
Comments

BlueNexus committed:
Fixed a compile error

BlueNexus committed:
Small fix

BlueNexus committed:
Small fix

BlueNexus committed:
Fixed assemblies eating parts

Hopefully, at least.

BlueNexus committed:
Added assembly handling for sights and stocks

Now for core.

BlueNexus committed:
Added a var for error handling

BlueNexus committed:
Refactored failure messages

BlueNexus committed:
Refactored failure messages and a minor fix

BlueNexus committed:
Glorious refactoring

Took out at least 25 lines. Procs are great.

BlueNexus committed:
Even more refactoring

BlueNexus committed:
Fixed the refactor

BlueNexus committed:
Fixed errors not returning

BlueNexus committed:
Cleaned up code, framework for decon

BlueNexus committed:
Removed unnecessary vars

BlueNexus committed:
Added a list

BlueNexus committed:
Added a part removal proc

BlueNexus committed:
Part removal update

BlueNexus committed:
Hopeful compile fix

BlueNexus committed:
Added attachment handling to disassembly

BlueNexus committed:
Added charge costs

BlueNexus committed:
Create base.dm

BlueNexus committed:
Base projectile file

BlueNexus committed:
Work-in-progress compile proc

BlueNexus committed:
Added casing handling to compile

BlueNexus committed:
Added heat level handling

BlueNexus committed:
Added a heat limit var for later use

BlueNexus committed:
Work-in-progress overheat handling

BlueNexus committed:
More heat handling

BlueNexus committed:
Cooldown handling

Not sure if this will work or not. We'll see.

BlueNexus committed:
Further handling

BlueNexus committed:
Create energy.dm

BlueNexus committed:
Extra quotation fix

BlueNexus committed:
Merge branch 'master' into ModularFirearms

BlueNexus committed:
Comments and punctuation

Very minor, but it's something to do while I figure out fire delays for energy weapons.
